Security/Principal/WindowsLoggedInIdentity.cs

```diff
@@ -36,23 +37,36 @@ public class WindowsLoggedInIdentity : IDisposable, IIdentity
 		/// The logon provider. This parameter can usually be left as the default. For more information, lookup more detail for the
 		/// dwLogonProvider parameter of the Windows LogonUser function.
 		/// </param>
-		public WindowsLoggedInIdentity(string userName, string domainName, string password, LogonUserType logonType = LogonUserType.LOGON32_LOGON_INTERACTIVE,
-			LogonUserProvider provider = LogonUserProvider.LOGON32_PROVIDER_DEFAULT)
+		public WindowsLoggedInIdentity(string userName, string domainName, string password, LogonUserType logonType = LogonUserType.LOGON32_LOGON_INTERACTIVE, LogonUserProvider provider = LogonUserProvider.LOGON32_PROVIDER_DEFAULT)
 		{
-			if (string.IsNullOrEmpty(userName)) throw new ArgumentNullException(nameof(userName));
-			if (string.IsNullOrEmpty(password)) throw new ArgumentNullException(nameof(password));
-			if (string.IsNullOrEmpty(domainName) && !userName.Contains("@")) throw new ArgumentNullException(nameof(domainName));
-			if (LogonUser(userName, domainName, password, logonType, provider, out hToken))
+			if (Extensions.StringHelper.IsNullOrWhiteSpace(userName))
+				throw new ArgumentNullException(nameof(userName));
+
+			if (Extensions.StringHelper.IsNullOrWhiteSpace(password))
+				throw new ArgumentNullException(nameof(password));
+
+			if (Extensions.StringHelper.IsNullOrWhiteSpace(domainName) && !userName.Contains("@"))
+				throw new ArgumentNullException(nameof(domainName));
+
+
+			var success = LogonUser(userName, domainName, password, logonType, provider, out hToken);
+
+			var lastError = Marshal.GetLastWin32Error();
```

Comment:
It is good practice that for every PInvoke call, Marshal.GetLastWin32Error(); is called immediately afterwards to ensure that the current error code (if any) is captured since .NET internally also makes/can PInvoke calls, possibly resulting in the wrong error code.

Reply:
A good read: https://devblogs.microsoft.com/oldnewthing/20150819-00/?p=91781

Reply:
The current code does this through the `Win32Exception` constructor.

Reply:
It would at least fix CA1404
https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1404?view=vs-2019

Reply:
If you look through Microsoft's P/Invoke use, you'll find that pattern over and over.